### PR TITLE
[TM-4843] Run Panel Integration

### PR DESCRIPTION
**Pull request by jlam-mpc**
Until further notice, all edit/run features are admin only.

## **Date Field Rules:**

MEET = Panel Meeting Date
* Required
* Populated at panel meeting creation
* Can be manually edited until date has passed

CUT = Preliminary Cutoff
* Required
* Populated at panel meeting creation
* Must be before MEET and ADD dates
* Can be manually edited until date has passed

ADD = Addendum Cutoff
* Required
* Populated at panel meeting creation
* Must be before MEET date
* Can be manually edited until date has passed

OFF = Preliminary Run Time
* Required
* Populated when run preliminary button is triggered
* Can only be populated once preliminary cutoff date has passed
* Cannot be manually edited at any time
* Can only be ran ONCE (user cannot click run preliminary again to fill input with new value)

OFFA = Addendum Run Time
* Required
* Populated when run addendum button is triggered
* Can only be populated once addendum cutoff date has passed
* Cannot be manually edited at any time
* Can only be ran ONCE (user cannot click run addendum again to fill input with new value)

POSS = Post Panel Started
* **NOTE: ANY NOTES ON THESE ARE PENDING CLARIFICATION FROM VICTORIA. Legacy FSBID does not have any rules nor does it enforce this field so it's up to TalentMap to define how this should best be used.**
* Optional?
* Previously assumed this would be automatically populated when the user opens the post panel tab for the first time (TM would send a put request to add that date to the panel meeting) but now I'm wondering if it's supposed to be left blank and filled in manually

POST = Post Panel Run Time
* Required
* Populated when run post panel button is triggered
* Can only be populated until POSS date exists and has passed. If POSS date is optional, then this requirement can be removed and POST can be populated with the run button at any point after addendum status
* Cannot be manually edited at any time
* Can only be ran ONCE (user cannot click run post panel again to fill input with new value)

COMP = Agenda Completed Time
* **NOTE: ANY NOTES ON THESE ARE PENDING CLARIFICATION FROM VICTORIA. Legacy FSBID does not have any rules nor does it enforce this field so it's up to TalentMap to define how this should best be used.**
* Optional?
* Populated manually by user entry and save
* Can only be populated once POST date exists (in other words, when Run Post Panel has been triggered)

## **Other Rules**

Run Official Preliminary Button
* Disabled until Preliminary Cutoff Date has passed
* Disabled again when Panel Meeting Date has passed
* There must be a subsequent panel meeting in Initiated status for the same meeting type so that "Not Ready" agenda items can roll forward.
* All agenda items must be in Approved, Ready, Not-Ready or Deferred - Proposed Position status.
    * In my testing, talentMap had created agenda items on a panel that were in status' that are only valid for Post Panel use (Deferred, Disapproved, Held, Move to ML/ID, Out of Order, PIP, and Withdrawn). 
    * Legacy FSBid does not present these states to CDOs or AOs when they are adding agenda items. If TalentMap does not limit from data entry, you will need to apply these rules or have difficulties running the Official Preliminary.
* Approved agenda items must be in the Off-Panel meeting category.
* If all rules are passed:
    * Not-Ready agenda items move to the next meeting.
    * All other items are numbered.
    * Meeting changes from initiated to Official Preliminary status.

Run Official Addendum Button
* Disabled until Addendum Cutoff Date has passed
* Disabled again when Panel Meeting Date has passed
* There must be a subsequent panel meeting in Initiated status for the same meeting type so that "Not Ready" agenda items can roll forward.
* Agenda items cannot be in Disapproved or Not-Ready status.
* If all rules are passed:
    * All items are numbered.
    * Meeting changes from Official Preliminary to Addendum status.

Run Post Panel Button
* No rules at this time (pending clarification again)

Post Panel Processing Tab
* Disabled until Panel Meeting page is fully populated (Official Addendum has been ran)

## **PR Info**

Dual Merge: 
- [API PR](https://github.com/MetaPhase-Consulting/State-TalentMAP-API/pull/850)
- [Mock PR](https://github.com/MetaPhase-Consulting/mock-fsbid/pull/407)

[TM-4843](https://metaphase.atlassian.net/browse/TM-4843)

[TM-4843]: https://metaphase.atlassian.net/browse/TM-4843?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Components/AdministratorPage/PanelAdmin/PanelMeetingAdmin.jsx [hunk 1]
```diff
@@ -281,7 +282,7 @@ const PanelMeetingAdmin = (props) => {
               <div className="date-picker-wrapper larger-date-picker">
                 <FA name="fa fa-calendar" onClick={() => openDatePicker()} />
                 <DatePicker
-                  disabled={disablePrelimRunTime}
+                  disabled
```
Comment (M00staff) [inline src/Components/AdministratorPage/PanelAdmin/PanelMeetingAdmin.jsx:285]: this Datpicker for prelim run time will always be disabled?

File: src/Components/AdministratorPage/PanelAdmin/PanelMeetingAdmin.jsx [hunk 2]
```diff
@@ -305,7 +306,7 @@ const PanelMeetingAdmin = (props) => {
               <div className="date-picker-wrapper larger-date-picker">
                 <FA name="fa fa-calendar" onClick={() => openDatePicker()} />
                 <DatePicker
-                  disabled={disableAddendumRunTime}
+                  disabled
```
Comment (M00staff) [inline src/Components/AdministratorPage/PanelAdmin/PanelMeetingAdmin.jsx:309]: always disabled?

File: src/actions/panelMeetingAdmin.js [hunk 3]
```diff
@@ -48,3 +54,101 @@ export function createPanelMeeting(props) {
...
+  }
+  return (dispatch) => {
+    if (cancelRunPanel) {
+      cancelRunPanel('cancel');
+    }
+    batch(() => {
+      dispatch(errored(false));
+    });
+    const ep = `/fsbid/admin/panel/run/${type}/${id}/`;
+    api().put(ep, {
```
Comment (M00staff) [inline src/actions/panelMeetingAdmin.js:133]: i think this would be a POST request

File: src/actions/panelMeetingAdmin.js [hunk 4]
```diff
@@ -48,3 +54,101 @@ export function createPanelMeeting(props) {
...
+      cancelToken: new CancelToken((c) => { cancelRunPanel = c; }),
```
Comment (M00staff) [inline src/actions/panelMeetingAdmin.js:134]: 👍

File: src/actions/postPanelProcessing.js [hunk 5]
```diff
@@ -1,11 +1,15 @@
 import { batch } from 'react-redux';
+import { CancelToken } from 'axios';
+// import Q from 'q';
```
Comment (M00staff) [inline src/actions/postPanelProcessing.js:3]: i think we can remove this - unless you wanna have the request stringified in the params?

File: src/sass/_positionContent.scss [hunk 6]
```diff
@@ -73,12 +73,15 @@
...
+
   .toggle-edit-mode-disabled {
     color: gray;
     cursor: not-allowed;
 
     &:hover {
-      div, span {
+
+      div,
+      span {
```
Comment (M00staff) [inline src/sass/_positionContent.scss:84]: nitpick - this seems like a weird linting thing?

File: src/sass/_positionContent.scss [hunk 7]
```diff
@@ -309,7 +320,8 @@
   .position-form--label-input-container-flex {
     display: flex;
 
-    select, input {
+    select,
+    input {
```
Comment (M00staff) [inline src/sass/_positionContent.scss:324]: doesn't show me squiggles - so its fine!

**Comment by M00staff** [comment]
just looking at the code - and everything looks OK - actions are all mapping up.

gonna go through the API next

**Review by M00staff — APPROVED** [review]
The PUT/POST stuff is the only thing that I think is worth changing ... though, if it's working, it might not need to be - your call. 

approving regardless.

File: src/Components/AdministratorPage/PanelAdmin/PanelMeetingAdmin.jsx (hunk 1 above)
Comment (jlam-mpc) [reply src/Components/AdministratorPage/PanelAdmin/PanelMeetingAdmin.jsx:285]: Yes: So for any runtime input WS let me me know that those cannot be manually edited/entered and can only be populated using the "Run ___" buttons, so they should all be disabled and view only.

File: src/Components/AdministratorPage/PanelAdmin/PanelMeetingAdmin.jsx (hunk 2 above)
Comment (jlam-mpc) [reply src/Components/AdministratorPage/PanelAdmin/PanelMeetingAdmin.jsx:309]: Same comment as my reply for run prelim!

File: src/sass/_positionContent.scss (hunk 6 above)
Comment (jlam-mpc) [reply src/sass/_positionContent.scss:84]: When I format my sass files any comma separated set of selectors automatically splits into new lines whether its 3+ or just 2. I thought everyones was formatting this way but if not I can revert this manually and check into my formatter for a bit.

File: src/actions/postPanelProcessing.js (hunk 5 above)
Comment (jlam-mpc) [reply src/actions/postPanelProcessing.js:3]: Oops I initially planned to do a bulk send for the edit calls but I decided against it and accidentally left this in. Will remove-

**Comment by scott062** [comment]
Just need to update with dev when you're ready

**Pull request merged**